### PR TITLE
Disable redirects for Ajax requests

### PR DESCRIPTION
**Pull request by HaoK**
cc @Tratcher @blowdart

### CODE REVIEW

**Comment by Tratcher** [comment]
#508. @PinpointTownes

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs [hunk 1]
```diff
@@ -42,6 +43,28 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
             return Task.FromResult(0);
         };
 
+        private static bool IsAjaxRequest(HttpRequest request)
+        {		
+            var query = request.Query;		
+            if (query != null)		
+            {		
+                if (query["X-Requested-With"] == "XMLHttpRequest")
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: These null checks shouldn't be needed. Worse case you can combine them into a single if.

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Indeed, collapsed into one return, I just lifted this code from the old stuff as is before :)

**Comment by PinpointTownes** [comment]
Nice.

Could we also remove these checks and move the default values to the options class? https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationMiddleware.cs#L65-L76

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs [hunk 2]
```diff
@@ -71,24 +94,73 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
...
+        }
 
         /// <summary>
         /// Implements the interface method by invoking the related delegate method.
         /// </summary>
         /// <param name="context">Contains information about the event</param>
-        public virtual Task RedirectToLogin(CookieRedirectContext context) => OnRedirect(context);
+        public virtual Task RedirectToLogin(CookieRedirectContext context)
+        {
+            if (IsAjaxRequest(context.Request))
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: This isn't the right place for a default, it's very difficult to override. It should go into OnRedirect.

Why isn't there a different OnRedirect* for each of these?

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: I believe you suggested to collapse them all into one in the last PR for this :)  They started all different

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Really? Then it was a terrible idea, nobody uses these directly. Maybe I meant that the same func could be assigned to each of them.

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: That said unless we had a central redirect place that took an optional UseThisStatusCodeForAjax parameter, its always going to be a bit meh with 4 different redirect events :/

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs [hunk 3]
```diff
@@ -71,24 +78,73 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
         /// Implements the interface method by invoking the related delegate method.
         /// </summary>
         /// <param name="context">Contains information about the event</param>
-        public virtual Task RedirectToLogout(CookieRedirectContext context) => OnRedirect(context);
+        public virtual Task RedirectToLogout(CookieRedirectContext context)
+        {
+            if (IsAjaxRequest(context.Request))
+            {
+                context.Response.Headers["Location"] = context.RedirectUri;
+                context.Response.StatusCode = 200;
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: No need to set the status code for logout.

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs [hunk 4]
```diff
@@ -71,24 +78,73 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
...
         /// Implements the interface method by invoking the related delegate method.
         /// </summary>
         /// <param name="context">Contains information about the event</param>
-        public virtual Task RedirectToReturnUrl(CookieRedirectContext context) => OnRedirect(context);
+        public virtual Task RedirectToReturnUrl(CookieRedirectContext context)
+        {
+            if (IsAjaxRequest(context.Request))
+            {
+                context.Response.Headers["Location"] = context.RedirectUri;
+                context.Response.StatusCode = 200;
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: No need to set the status code here.

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs (hunk 2 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Okay I'll move these to a new OnXyz for each redirect.  Still seems a bit meh that each one will need to be overriden

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Yeah, but without individual On's you still don't know which redirect is happening.

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Ideally, why not moving these default "behaviors" to the handler itself, just like the other events classes?

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: You could do that if you used BaseControlContext.

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Sure. `CookieRedirectContext` seems a good candidate for that :smile:

**Comment by Tratcher** [comment]
@PinpointTownes Removing those checks would break it unless we also added a "don't redirect if these are null" logic. Using the events is a better way to take control.

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs (hunk 3 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: If we don't set this, the status code appears to stay 401 for the logout test

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs (hunk 2 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Eh cookies is the only one that isn't currently using control context, I thought it was intentional, but I can't remember anymore honestly :)

**Comment by PinpointTownes** [comment]
@Tratcher IMHO, having to use the events to simply avoid the default redirects is really annoying. There are definitely cases where you don't want to automagically replace 401 responses by 302 redirects.

**Comment by HaoK** [comment]
Do we just want a global options switch for cookies to turn off all redirects?  Options.DoNotRedirect = true?

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs (hunk 3 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: why, who set it?

**Comment by Tratcher** [comment]
We already have it, Options.AutomaticChallenge = false.

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs [hunk 5]
```diff
@@ -42,6 +43,12 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
             return Task.FromResult(0);
         };
 
+        private static bool IsAjaxRequest(HttpRequest request)
+        {
+            return request?.Query["X-Requested-With"] == "XMLHttpRequest" ||
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Curious: why would `request` be null?

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Don't bother with null checks here, these are core APIs.

File: src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs (hunk 3 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs]: Event was wired up to login not return url, least the tests are catching stuff :)  Fixed and updated

**Comment by Tratcher** [comment]
:shipit:

**Comment by HaoK** [comment]
1d2c6ba1223748df02a879622e8deb6ac23db5b7

**Pull request closed without merging**